src/lib.rs

```diff
@@ -506,43 +506,71 @@ mod opassign {
 
     impl<T: Clone + Integer + NumAssign> AddAssign for Ratio<T> {
         fn add_assign(&mut self, other: Ratio<T>) {
-            self.numer *= other.denom.clone();
-            self.numer += self.denom.clone() * other.numer;
-            self.denom *= other.denom;
+            if self.denom == other.denom {
+                self.numer += other.numer
+            } else {
+                let lcm = self.denom.lcm(&other.denom.clone());
+                let lhs_numer = self.numer.clone() * (lcm.clone() / self.denom.clone());
+                let rhs_numer = other.numer * (lcm.clone() / other.denom);
+                self.numer = lhs_numer + rhs_numer;
+                self.denom = lcm;
+            }
             self.reduce();
         }
     }
 
+    // (a/b) / (c/d) = (a/gcd_ac)*(d/gcd_bd) / ((c/gcd_ac)*(b/gcd_bd))
     impl<T: Clone + Integer + NumAssign> DivAssign for Ratio<T> {
         fn div_assign(&mut self, other: Ratio<T>) {
-            self.numer *= other.denom;
-            self.denom *= other.numer;
-            self.reduce();
+            let gcd_ac = self.numer.gcd(&other.numer.clone());
+            let gcd_bd = self.denom.gcd(&other.denom.clone());
+            self.numer /= gcd_ac.clone();
+            self.numer *= other.denom / gcd_bd.clone();
+            self.denom /= gcd_bd.clone();
+            self.denom *= other.numer / gcd_ac.clone();
```

Comment:
The `gcd` calls don't need `clone()`, and the final use of `gcd_bd` and `gcd_ac` can be used directly too.